tools/lsp/ui/main.slint

```diff
@@ -0,0 +1,49 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-1.1 OR LicenseRef-Slint-commercial
+
+// cSpell: ignore Heade
+
+import { ScrollView, Button, HorizontalBox } from "std-widgets.slint";
+import { HeaderBar } from "header-bar.slint";
+import { Spinner } from "spinner.slint";
+
+export component PreviewUi inherits Window {
+    in property<component-factory> preview-area <=> i-preview-area-container.component-factory;
+    callback design-mode-changed(/* enabled */ bool);
+
+    in property <bool> is-busy;
+
+    min-width: min(i-scroll-view.viewport-width, 100px);
+    min-height: min(i-scroll-view.viewport-height, 100px);
+
+    VerticalLayout {
+        HeaderBar {
+            i-pick-button := Button {
+                icon: self.checked ? @image-url("../assets/colorize-light.svg") :
+                    @image-url("../assets/colorize.svg");
```

Comment:
It seems that this should depend on the light-mode or not. 
This just relies on the internal of the fluent style.  And it will break with qt and cupertino style, i think.

I don't have a solution for that. But this is not good.  I wonder if it wouldn't make sense to have a colorfull icon instead.  Or just an icon with contrast included so it looks alright everywhere

Reply:
But this behavior is not alright.  We should have an icon that works both in dark and light styles.

> Ideally this would be an icon font [...]. AFAICT that is not an option at this time though.

Why is it not an option?

Reply:
I can prepare something tomorrow that we have an icon that works also for the dark styles.

Reply:
> You can e.g. not change the font used on our standard button.

Indeed, that's the problem.

Otherwise for a Text you would just import the font, and use it in font-family. And then use `text: "\u{xxx}"`.
But without setting the font to a button, this is indeed impossible :-(

I wonder if we could register global "fallback" font maybe that use the private unicode range so that we can use that as text and if the glyph is in the fallback it'd use that.

Anyway, i'm a bit off topic here.

Reply:
Easily for this topic I would use svg icons. And do it so it looks good to all styles and dark and light.

For coop I'm using svg icons, but use internal colorize to set the color of the icons depending on the palette. It would be possible to do also for the std widgets but this would cause trouble with colorful images/svgs. We could also add  small Icon set to each style that is available in public api. 

I think for the std-widgets there are different solutions to solve the icon topic, but we need to discuss it. So as mentioned I can contribute a solution tomorrow that fix our problem for now for this situation.

Reply:
It is not that easy that I thought. What we could to general is to introduce an `ButtonBase` that can also profit users from that want to create buttons with custom content within:


### Example for the cupertino `ButtonBase`

```slint
export component ButtonBase {
    callback clicked;

    in property <bool> primary;
    in property <bool> enabled <=> i-touch-area.enabled;
    in property <bool> checkable;
    out property <bool> has-focus: i-focus-scope.has-focus;
    out property <bool> pressed: self.enabled && i-touch-area.pressed;
    out property <brush> background: primary && root.enabled ? Palette.accent : Palette.surface;
    out property <brush> foreground: primary && root.enabled ? Palette.on-surface : Palette.foreground;
    in-out property <bool> checked;

    min-width: max(20px, i-layout.min-width);
    min-height: max(20px, i-layout.min-height);
    horizontal-stretch: 0;
    vertical-stretch: 0;
    accessible-role: button;

    FocusBorder {
        x: (parent.width - self.width) / 2;
        y: (parent.height - self.height) / 2;
        width: parent.width + 6px;
        height: parent.height + 6px;
        border-radius: 8px;
        has-focus: root.has-focus;
    }

    if (root.primary && root.enabled) : Rectangle {
        drop-shadow-blur: 3px;
        drop-shadow-color: #00000066;
        drop-shadow-offset-y: 0.5px;
        border-radius: 5px;
        background: root.background;

        Rectangle {
            drop-shadow-blur: 2px;
            drop-shadow-color: #00000026;
            drop-shadow-offset-y: 1px;
            border-radius: parent.border-radius;
            background: root.background;
        }

        Rectangle {
            drop-shadow-blur: 1px;
            drop-shadow-color: #00000026;
            drop-shadow-offset-y: 0.5px;
            border-radius: parent.border-radius;
            background: root.background;
        }

        Rectangle {
            border-radius: parent.border-radius;
            background: Palette.dimmer;
            opacity: 0.17;
        }
    }

    if (!root.primary || !root.enabled) : Rectangle {
        drop-shadow-blur: 0.25px;
        drop-shadow-color: #00000066;
        drop-shadow-offset-y: 0.25px;
        border-radius: 5px;
        background: root.background;

        Rectangle {
            drop-shadow-blur: 1px;
            drop-shadow-color: #00000026;
            drop-shadow-offset-y: 1px;
            border-radius: parent.border-radius;
            background: root.background;
            border-width: 1px;
            border-color: Palette.decent-border;
            opacity: root.enabled ? 1 : 0.5;
        }
    }

    i-layout := HorizontalLayout {
        padding-left: 8px;
        padding-right: 8px;
        padding-top: 4px;
        padding-bottom: 4px;
        spacing: 4px;
        alignment: center;

        @children
    }

    i-touch-area := TouchArea {
        clicked => {
            if (root.checkable) {
                root.checked = !root.checked;
            }
            root.clicked();
        }
    }

    i-focus-scope := FocusScope {
        x: 0;
        width: 0; // Do not react on clicks
        enabled <=> root.enabled;

        key-pressed(event) => {
            if (event.text == " " || event.text == "\n") {
                i-touch-area.clicked();
                 return accept;
            }

            return reject;
        }
    }

    states [
        disabled when !i-touch-area.enabled : {
            root.foreground: Palette.foreground-secondary;
            root.background: Palette.surface-quaternary;
        }
        pressed when root.pressed : {
            root.background: root.primary ? Palette.accent-secondary : Palette.surface-secondary;
        }
        checked when root.checked : {
            root.foreground: Palette.accent-secondary;
        }
    ]

    animate background { duration: 150ms; }
}
```

And then in our use case:

```slint
import { ButtonBase } from "std-widgets.slint`

export component IconButton {
     in property <image> icon <=> i-icon.source;
     
     i-icon := Image {
          colorize: root.foreground;
      }
}
```

We could use the `ButtonBase` also as base for our buttons.

Reply:
But @ogoffart could be difficult to implement this for the qt style right? Otherwise I think it make sense for this use case and also some users would be profit, because they can define buttons with custom content.